investigate flacky test reactor.core.publisher.tck.FluxBlackboxProcessorVerification > required_spec313_cancelMustMakeThePublisherEventuallyDropAllReferencesToTheSubscriber

https://github.com/reactor/reactor-core/actions/runs/7172808161/job/19530791917#step:8:1549

```
java.lang.AssertionError: Async error during test execution: Scheduler unavailable
        at org.testng.Assert.fail(Assert.java:98)
        at org.reactivestreams.tck.TestEnvironment.verifyNoAsyncErrorsNoDelay(TestEnvironment.java:367)
        at org.reactivestreams.tck.PublisherVerification$27.run(PublisherVerification.java:1021)
        at org.reactivestreams.tck.PublisherVerification.activePublisherTest(PublisherVerification.java:1135)
        at org.reactivestreams.tck.PublisherVerification.required_spec313_cancelMustMakeThePublisherEventuallyDropAllReferencesToTheSubscriber(PublisherVerification.java:1007)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.testng.internal.invokers.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:136)
        at org.testng.internal.invokers.TestInvoker.invokeMethod(TestInvoker.java:658)
        at org.testng.internal.invokers.TestInvoker.invokeTestMethod(TestInvoker.java:219)
        at org.testng.internal.invokers.MethodRunner.runInSequence(MethodRunner.java:50)
        at org.testng.internal.invokers.TestInvoker$MethodInvocationAgent.invoke(TestInvoker.java:923)
        at org.testng.internal.invokers.TestInvoker.invokeTestMethods(TestInvoker.java:192)
        at org.testng.internal.invokers.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:146)
        at org.testng.internal.invokers.TestMethodWorker.run(TestMethodWorker.java:128)
        at java.util.ArrayList.forEach(ArrayList.java:1259)
        at org.testng.TestRunner.privateRun(TestRunner.java:808)
        at org.testng.TestRunner.run(TestRunner.java:603)
        at org.testng.SuiteRunner.runTest(SuiteRunner.java:429)
        at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:423)
        at org.testng.SuiteRunner.privateRun(SuiteRunner.java:383)
        at org.testng.SuiteRunner.run(SuiteRunner.java:326)
        at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
        at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:95)
        at org.testng.TestNG.runSuitesSequentially(TestNG.java:1249)
        at org.testng.TestNG.runSuitesLocally(TestNG.java:1169)
        at org.testng.TestNG.runSuites(TestNG.java:1092)
        at org.testng.TestNG.run(TestNG.java:1060)
        at org.gradle.api.internal.tasks.testing.testng.TestNGTestClassProcessor.runTests(TestNGTestClassProcessor.java:148)
        at org.gradle.api.internal.tasks.testing.testng.TestNGTestClassProcessor.stop(TestNGTestClassProcessor.java:95)
        at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.stop(SuiteTestClassProcessor.java:62)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
        at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
        at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
        at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
        at com.sun.proxy.$Proxy2.stop(Unknown Source)
        at org.gradle.api.internal.tasks.testing.worker.TestWorker$3.run(TestWorker.java:193)
        at org.gradle.api.internal.tasks.testing.worker.TestWorker.executeAndMaintainThreadName(TestWorker.java:129)
        at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:100)
        at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:60)
        at org.gradle.process.internal.worker.child.ActionExecutionWorker.execute(ActionExecutionWorker.java:56)
        at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:113)
        at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:65)
        at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
        at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)

        Caused by:
        reactor.core.Exceptions$ReactorRejectedExecutionException: Scheduler unavailable
            at reactor.core.Exceptions.failWithRejected(Exceptions.java:283)
            at reactor.core.publisher.Operators.onRejectedExecution(Operators.java:1075)
            at reactor.core.publisher.FluxPublishOn$PublishOnSubscriber.trySchedule(FluxPublishOn.java:327)
            at reactor.core.publisher.FluxPublishOn$PublishOnSubscriber.onNext(FluxPublishOn.java:237)
            at reactor.core.publisher.FluxFlatMap$FlatMapMain.tryEmit(FluxFlatMap.java:547)
            at reactor.core.publisher.FluxFlatMap$FlatMapInner.onNext(FluxFlatMap.java:988)
            at reactor.core.publisher.FluxFlatMap$FlatMapMain.tryEmit(FluxFlatMap.java:547)
            at reactor.core.publisher.FluxFlatMap$FlatMapInner.onNext(FluxFlatMap.java:988)
            at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1865)
            at reactor.core.publisher.FluxZip$ZipSingleCoordinator.next(FluxZip.java:514)
            at reactor.core.publisher.FluxZip$ZipSingleSubscriber.onNext(FluxZip.java:650)
            at reactor.core.publisher.FluxArray$ArraySubscription.fastPath(FluxArray.java:171)
            at reactor.core.publisher.FluxArray$ArraySubscription.request(FluxArray.java:96)
            at reactor.core.publisher.FluxZip$ZipSingleSubscriber.onSubscribe(FluxZip.java:633)
            at reactor.core.publisher.FluxArray.subscribe(FluxArray.java:53)
            at reactor.core.publisher.FluxArray.subscribe(FluxArray.java:59)
            at reactor.core.publisher.Flux.subscribe(Flux.java:8777)
            at reactor.core.publisher.FluxZip$ZipSingleCoordinator.subscribe(FluxZip.java:488)
            at reactor.core.publisher.FluxZip.handleBoth(FluxZip.java:306)
            at reactor.core.publisher.FluxZip.handleArrayMode(FluxZip.java:273)
            at reactor.core.publisher.FluxZip.subscribe(FluxZip.java:137)
            at reactor.core.publisher.Flux.subscribe(Flux.java:8777)
            at reactor.core.publisher.FluxFlatMap$FlatMapMain.onNext(FluxFlatMap.java:430)
            at reactor.core.publisher.FluxFlatMap$FlatMapMain.drainLoop(FluxFlatMap.java:716)
            at reactor.core.publisher.FluxFlatMap$FlatMapMain.drain(FluxFlatMap.java:592)
            at reactor.core.publisher.FluxFlatMap$FlatMapInner.onSubscribe(FluxFlatMap.java:959)
            at reactor.core.publisher.FluxIterable.subscribe(FluxIterable.java:201)
            at reactor.core.publisher.FluxIterable.subscribe(FluxIterable.java:83)
            at reactor.core.publisher.Flux.subscribe(Flux.java:8777)
            at reactor.core.publisher.FluxFlatMap$FlatMapMain.onNext(FluxFlatMap.java:430)
            at reactor.core.publisher.SerializedSubscriber.onNext(SerializedSubscriber.java:99)
            at reactor.core.publisher.FluxBufferTimeout$BufferTimeoutSubscriber.flushCallback(FluxBufferTimeout.java:654)
            at reactor.core.publisher.FluxBufferTimeout$BufferTimeoutSubscriber.checkedComplete(FluxBufferTimeout.java:731)
            at reactor.core.publisher.FluxBufferTimeout$BufferTimeoutSubscriber.onComplete(FluxBufferTimeout.java:788)
            at reactor.core.publisher.FluxMap$MapSubscriber.onComplete(FluxMap.java:144)
            at reactor.core.publisher.FluxFilter$FilterSubscriber.onComplete(FluxFilter.java:166)
            at reactor.core.publisher.FluxMap$MapConditionalSubscriber.onComplete(FluxMap.java:275)
            at reactor.core.publisher.FluxScan$ScanSubscriber.onComplete(FluxScan.java:132)
            at reactor.core.publisher.FluxPeekFuseable$PeekFuseableSubscriber.onComplete(FluxPeekFuseable.java:277)
            at reactor.core.publisher.FluxPublishOn$PublishOnSubscriber.doComplete(FluxPublishOn.java:505)
            at reactor.core.publisher.FluxPublishOn$PublishOnSubscriber.checkTerminated(FluxPublishOn.java:552)
            at reactor.core.publisher.FluxPublishOn$PublishOnSubscriber.runAsync(FluxPublishOn.java:432)
            at reactor.core.publisher.FluxPublishOn$PublishOnSubscriber.run(FluxPublishOn.java:527)
            at reactor.core.scheduler.WorkerTask.call(WorkerTask.java:84)
            at reactor.core.scheduler.WorkerTask.call(WorkerTask.java:37)
            at java.util.concurrent.FutureTask.run(FutureTask.java:266)
            at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
            at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
            at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
            at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
            at java.lang.Thread.run(Thread.java:750)

            Caused by:
            reactor.core.Exceptions$StaticRejectedExecutionException: Scheduler unavailable
```

--- a/reactor-core/src/jcstress/java/reactor/core/publisher/FluxPublishOnStressTest.java
+++ b/reactor-core/src/jcstress/java/reactor/core/publisher/FluxPublishOnStressTest.java
@@ -1,3 +1,19 @@
+/*
+ * Copyright (c) 2023 VMware Inc. or its affiliates, All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
 package reactor.core.publisher;
 
 import java.util.concurrent.ForkJoinPool;

--- a/reactor-core/src/jcstress/java/reactor/core/publisher/FluxPublishOnStressTest.java
+++ b/reactor-core/src/jcstress/java/reactor/core/publisher/FluxPublishOnStressTest.java
@@ -1,0 +1,96 @@
+package reactor.core.publisher;
+
+import java.util.concurrent.ForkJoinPool;
+
+import org.openjdk.jcstress.annotations.Actor;
+import org.openjdk.jcstress.annotations.Arbiter;
+import org.openjdk.jcstress.annotations.JCStressTest;
+import org.openjdk.jcstress.annotations.Outcome;
+import org.openjdk.jcstress.annotations.State;
+import org.openjdk.jcstress.infra.results.II_Result;
+import reactor.core.scheduler.Schedulers;
+import reactor.core.scheduler.Scheduler;
+import reactor.util.concurrent.Queues;
+
+import static org.openjdk.jcstress.annotations.Expect.ACCEPTABLE;
+
+public class FluxPublishOnStressTest {
+
+	@JCStressTest
+	@Outcome(id = {"0, 1", "0, 0"}, expect = ACCEPTABLE, desc = "no errors propagated after cancellation because of disposed worker")
+	@State
+	public static class FluxPublishOnOnNextAndCancelRaceStressTest {
+		final StressSubscription<Integer> upstream = new StressSubscription<>(null);
+		final StressSubscriber<Integer> downstream = new StressSubscriber<>();
+		final Scheduler scheduler =
+				Schedulers.fromExecutorService(ForkJoinPool.commonPool());
+
+		final FluxPublishOn.PublishOnSubscriber<Integer> publishOnSubscriber =
+				new FluxPublishOn.PublishOnSubscriber<>(downstream,
+						scheduler,
+						scheduler.createWorker(), true, 32, 12, Queues.get(32));
+
+
+		{
+			publishOnSubscriber.onSubscribe(upstream);
+		}
+
+		@Actor
+		public void produce() {
+			publishOnSubscriber.onNext(1);
+			publishOnSubscriber.onNext(2);
+			publishOnSubscriber.onNext(3);
+			publishOnSubscriber.onNext(4);
+		}
+
+		@Actor
+		public void cancel() {
+			publishOnSubscriber.cancel();
+		}
+
+		@Arbiter
+		public void arbiter(II_Result result) {
+			result.r1 = downstream.onErrorCalls.get();
+			result.r2 = downstream.droppedErrors.size();
+		}
+	}
+
+	@JCStressTest
+	@Outcome(id = {"0, 1", "0, 0"}, expect = ACCEPTABLE, desc = "no errors propagated after cancellation because of disposed worker")
+	@State
+	public static class FluxPublishOnConditionalOnNextAndCancelRaceStressTest {
+		final StressSubscription<Integer> upstream = new StressSubscription<>(null);
+		final ConditionalStressSubscriber<Integer> downstream = new ConditionalStressSubscriber<>();
+		final Scheduler scheduler =
+				Schedulers.fromExecutorService(ForkJoinPool.commonPool());
+
+		final FluxPublishOn.PublishOnConditionalSubscriber<Integer> publishOnSubscriber =
+				new FluxPublishOn.PublishOnConditionalSubscriber<>(downstream,
+						scheduler,
+						scheduler.createWorker(), true, 32, 12, Queues.get(32));
+
+
+		{
+			publishOnSubscriber.onSubscribe(upstream);
+		}
+
+		@Actor
+		public void produce() {
+			publishOnSubscriber.onNext(1);
+			publishOnSubscriber.onNext(2);
+			publishOnSubscriber.onNext(3);
+			publishOnSubscriber.onNext(4);
+		}
+
+		@Actor
+		public void cancel() {
+			publishOnSubscriber.cancel();
+		}
+
+		@Arbiter
+		public void arbiter(II_Result result) {
+			result.r1 = downstream.onErrorCalls.get();
+			result.r2 = downstream.droppedErrors.size();
+		}
+	}
+}

--- a/reactor-core/src/main/java/reactor/core/publisher/FluxPublishOn.java
+++ b/reactor-core/src/main/java/reactor/core/publisher/FluxPublishOn.java
@@ -324,6 +324,12 @@ final class FluxPublishOn<T> extends InternalFluxOperator<T, T> implements Fusea
 					// In all other modes we are free to discard queue immediately since there is no racing on pooling
 					Operators.onDiscardQueueWithClear(queue, actual.currentContext(), null);
 				}
+
+				if (cancelled) {
+					Operators.onErrorDropped(ree, actual.currentContext());
+					return;
+				}
+
 				actual.onError(Operators.onRejectedExecution(ree, subscription, suppressed, dataSignal,
 						actual.currentContext()));
 			}
@@ -884,6 +890,12 @@ final class FluxPublishOn<T> extends InternalFluxOperator<T, T> implements Fusea
 					// In all other modes we are free to discard queue immediately since there is no racing on pooling
 					Operators.onDiscardQueueWithClear(queue, actual.currentContext(), null);
 				}
+
+				if (cancelled) {
+					Operators.onErrorDropped(ree, actual.currentContext());
+					return;
+				}
+
 				actual.onError(Operators.onRejectedExecution(ree, subscription, suppressed, dataSignal,
 						actual.currentContext()));
 			}

--- a/reactor-core/src/main/java/reactor/core/publisher/FluxPublishOn.java
+++ b/reactor-core/src/main/java/reactor/core/publisher/FluxPublishOn.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2016-2021 VMware Inc. or its affiliates, All Rights Reserved.
+ * Copyright (c) 2016-2023 VMware Inc. or its affiliates, All Rights Reserved.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.

